File: rpi_ros2_ws/src/gui/gui/static/shared/protocol.js
const GuiProtocol = Object.freeze({
    websocketPath: "/websocket",
    websocketSubprotocol: "protocolOne",
    fields: Object.freeze({
        type: "type",
        data: "data",
        topicName: "topic_name",
        msg: "msg",
        actionName: "action_name",
        target: "target",
        action: "action",
        group: "group",
        params: "params",
    }),
    types: Object.freeze({
        action: "action",
        topic: "topic",
        process: "process",
        controller: "controller",
    }),
    actions: Object.freeze({
        initializeAllThrusters: "initialize_all_thrusters",
        flashStm32: "flash_stm32",
        setSupervisorSimulationMode: "set_supervisor_simulation_mode",
        setSupervisorManualMode: "set_supervisor_manual_mode",
        moveToPoint: "move_to_point",
        cancelMoveToPoint: "cancel_move_to_point",
    }),
    processTargets: Object.freeze({}),
    processActions: Object.freeze({
        start: "start",
        stop: "stop",
        kill: "kill",
    }),
    controllerGroups: Object.freeze({
        bottomCameraPidFbc: "bottom_camera_pid_fbc",
        depthControl: "depth_control",
    }),
    controllerActions: Object.freeze({
        enable: "enable",
        disable: "disable",
        reset: "reset",
        setPidParams: "set_pid_params",
    }),
    topics: Object.freeze({
        killed: "sensors/killed",
        depthM: "sensors/depth_m",
        stm32Log: "diagnostics/stm32/log",
        systemManagerMode: "system_manager/mode",
        systemManagerStatus: "system_manager/status",
        bottomCameraPidXReferencePx: "control/pid/bottom_camera/x/reference_px",
        bottomCameraPidYReferencePx: "control/pid/bottom_camera/y/reference_px",
        bottomCameraYawTargetRad: "control/targets/bottom_camera/yaw_rad",
        bottomCameraPidYawReferenceRad: "control/pid/bottom_camera/yaw/reference_rad",
        bottomCameraPidXFeedbackPx: "control/pid/bottom_camera/x/feedback_px",
        bottomCameraPidYFeedbackPx: "control/pid/bottom_camera/y/feedback_px",
        bottomCameraPidYawFeedbackRad: "state/bottom_camera/yaw_rad",
        thrustersPwmUs: "thrusters/pwm_us",
        thrustersEnabled: "thrusters/enabled",
        electromagnetEnabled: "actuators/electromagnet/enabled",
        wrenchCommand: "control/wrench_command",
        targetDepthM: "control/targets/depth_m",
        flashStm32Status: "flash_stm32_status",
        moveToPointStatus: "control/targets/move_to_point/gui_status",
    }),
    makeWebsocketUrl(hostname) {
        return "ws://" + hostname + this.websocketPath;
    },
    makeTopicMessage(topicName, msg) {
        return {
            type: this.types.topic,
            data: {
                topic_name: topicName,
                msg: msg,
            },
        };
    },
    makeActionMessage(actionName, data = {}) {
        return {
            type: this.types.action,
            data: {
                ...data,
                action_name: actionName,
            },
        };
    },
    makeProcessMessage(target, action) {
        return {
            type: this.types.process,
            data: {
                target: target,
                action: action,
            },
        };
    },
    makeControllerMessage(group, action, data = {}) {
        return {
            type: this.types.controller,
            data: {
                ...data,
                group: group,
                action: action,
            },
        };
    },
});

File: rpi_ros2_ws/src/gui/gui/backend/protocol.py
"""Shared websocket protocol constants for the GUI package."""

WEBSOCKET_PATH = "/websocket"
WEBSOCKET_SUBPROTOCOL = "protocolOne"

FIELD_TYPE = "type"
FIELD_DATA = "data"
FIELD_TOPIC_NAME = "topic_name"
FIELD_MSG = "msg"
FIELD_ACTION_NAME = "action_name"
FIELD_TARGET = "target"
FIELD_ACTION = "action"
FIELD_GROUP = "group"
FIELD_PARAMS = "params"

TYPE_ACTION = "action"
TYPE_TOPIC = "topic"
TYPE_PROCESS = "process"
TYPE_CONTROLLER = "controller"
MESSAGE_TYPES = (TYPE_ACTION, TYPE_TOPIC, TYPE_PROCESS, TYPE_CONTROLLER)

ACTION_INITIALIZE_ALL_THRUSTERS = "initialize_all_thrusters"
ACTION_FLASH_STM32 = "flash_stm32"
ACTION_SET_SUPERVISOR_SIMULATION_MODE = "set_supervisor_simulation_mode"
ACTION_MOVE_TO_POINT = "move_to_point"
ACTION_CANCEL_MOVE_TO_POINT = "cancel_move_to_point"

PROCESS_BOTTOM_CAMERA_PID_FBC_LAUNCH = "bottom_camera_pid_fbc_launch"
PROCESS_DEPTH_CONTROL_LAUNCH = "depth_control_launch"

CONTROLLER_GROUP_BOTTOM_CAMERA_PID_FBC = "bottom_camera_pid_fbc"
CONTROLLER_GROUP_DEPTH_CONTROL = "depth_control"

CONTROLLER_ACTION_ENABLE = "enable"
CONTROLLER_ACTION_DISABLE = "disable"
CONTROLLER_ACTION_RESET = "reset"
CONTROLLER_ACTION_SET_PID_PARAMS = "set_pid_params"

PROCESS_ACTION_START = "start"
PROCESS_ACTION_STOP = "stop"
PROCESS_ACTION_KILL = "kill"

SUPERVISOR_SERVICE_BOTTOM_CAMERA_HOLD = "bottom_camera_hold"
SUPERVISOR_SERVICE_DEPTH_HOLD = "depth_hold"
SUPERVISOR_SERVICE_DISABLE_BOTTOM_CAMERA_HOLD = "disable_bottom_camera_hold"
SUPERVISOR_SERVICE_DISABLE_DEPTH_HOLD = "disable_depth_hold"
SUPERVISOR_SERVICE_RESET_CONTROLLERS = "reset_controllers"
SUPERVISOR_SERVICE_SAFE_DISABLED = "safe_disabled"
SUPERVISOR_SERVICE_MANUAL = "manual"

TOPIC_KILLED = "sensors/killed"
TOPIC_DEPTH_M = "sensors/depth_m"
TOPIC_STM32_LOG = "diagnostics/stm32/log"
TOPIC_SYSTEM_MANAGER_MODE = "system_manager/mode"
TOPIC_SYSTEM_MANAGER_STATUS = "system_manager/status"
TOPIC_BOTTOM_CAMERA_PID_X_REFERENCE_PX = "control/pid/bottom_camera/x/reference_px"
TOPIC_BOTTOM_CAMERA_PID_Y_REFERENCE_PX = "control/pid/bottom_camera/y/reference_px"
TOPIC_BOTTOM_CAMERA_PID_X_FEEDBACK_PX = "control/pid/bottom_camera/x/feedback_px"
TOPIC_BOTTOM_CAMERA_PID_Y_FEEDBACK_PX = "control/pid/bottom_camera/y/feedback_px"
TOPIC_THRUSTERS_PWM_US = "thrusters/pwm_us"
TOPIC_THRUSTERS_ENABLED = "thrusters/enabled"
TOPIC_ELECTROMAGNET_ENABLED = "actuators/electromagnet/enabled"
TOPIC_WRENCH_COMMAND = "control/wrench_command"
TOPIC_TARGET_DEPTH_M = "control/targets/depth_m"
TOPIC_FLASH_STM32_STATUS = "flash_stm32_status"
TOPIC_MOVE_TO_POINT_STATUS = "control/targets/move_to_point/gui_status"

MOVE_TO_POINT_ACTION_NAME = "control/targets/move_to_point"


def topic_payload(topic_name, msg):
    """Build a websocket topic payload."""
    return {
        FIELD_TYPE: TYPE_TOPIC,
        FIELD_DATA: {
            FIELD_TOPIC_NAME: topic_name,
            FIELD_MSG: msg,
        },
    }
